Bring `http_network_or_cache_fetch` closer to the spec

simonwuelker opened the pull request:
This adds spec steps for the [ `http_network_or_cache_fetch`](https://fetch.spec.whatwg.org#http-network-or-cache-fetch) algorithm and implements the missing [`append_a_request_origin_header`](https://fetch.spec.whatwg.org/#append-a-request-origin-header) method, fixing a few WPT subtests in the process.

Draft because I would like to get https://github.com/whatwg/fetch/pull/1776 merged first and update the spec steps accordingly, but apart from that this is ready for review.

[try run](https://github.com/simonwuelker/servo/actions/runs/11013561747)

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] There are tests for these changes

simonwuelker committed:
Document "fetch" method

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
fix test-tidy errors

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1132,35 +1126,47 @@ async fn http_network_or_cache_fetch(
-            _ => None,
-        },
+    // Step 8.5 Let contentLength be httpRequest’s body’s length, if httpRequest’s body is non-null;
+    //          otherwise null.
```
Ah you're doing this indentation on purpose? Usually we don't, but why not(if test-tidy passes).

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1132,35 +1126,47 @@ async fn http_network_or_cache_fetch(
+    }
+
+    // Step 8.8 If contentLength is non-null, then set contentLengthHeaderValue to contentLength,
+    //          serialized and isomorphic encoded.
```
I don't think we're doing the "serialized and isomorphic encoded", TODO?

simonwuelker replied:
We don't need to, because we insert the header via `.typed_insert(ContentLength(content_length_header_value));`, which takes care of serialization. I guess i should add a note about that

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1132,35 +1126,47 @@ async fn http_network_or_cache_fetch(
-    // TODO Keepalive requires request's client object's fetch group
+    // Step 8.10 If contentLength is non-null and httpRequest’s keepalive is true, then:
+    if content_length.is_some() && http_request.keep_alive {
+        // TODO Keepalive requires request's client object's fetch group
```
Would be nice to have an issue for the sub-steps.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1309,8 +1309,7 @@ async fn http_network_or_cache_fetch(
 
-    // Step 5.18
-    // TODO If there’s a proxy-authentication entry, use it as appropriate.
+    // FIXME Step 8.22 If there’s a proxy-authentication entry, use it as appropriate.
```
TODO

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1508,88 +1514,29 @@ async fn http_network_or_cache_fetch(
+                // Step 10.5.2 Store httpRequest and forwardResponse in httpCache, as per the
+                //             "Storing Responses in Caches" chapter of HTTP Caching.
+                // TODO: The spec doesn't explicitly tell us to check for CacheMode::NoStore.
+                //       Is this correct?
```
Not the fetch spec, but the http cache spec does: https://httpwg.org/specs/rfc9111.html#response.cacheability

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1347,11 +1346,15 @@ async fn http_network_or_cache_fetch(
         }
 
-        // Step 5.19
+        // Step 8.23 Set httpCache to the result of determining the HTTP cache partition, given httpRequest.
```
This is a TODO, as we do not partition the cache.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1599,20 +1546,24 @@ async fn http_network_or_cache_fetch(
-    // Step 10
+    // FIXME: Step 11. Set response’s URL list to a clone of httpRequest’s URL list.
+    // FIXME: Step 12. If httpRequest’s header list contains `Range`, then set response’s range-requested flag.
+    // FIXME: Step 13 Set response’s request-includes-credentials to includeCredentials.
```
TODO if not implemented please.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1508,88 +1514,29 @@ async fn http_network_or_cache_fetch(
+    let mut response = response.unwrap();
 
+    // FIXME: The spec doesn't tell us to do this *here*, but if we don't do it then
+    //        tests fail. Where should we do it instead?
```
Looks like it's step 5 of HTTP Fetch, so just after this algo finishes: https://fetch.spec.whatwg.org/#ref-for-cross-origin-resource-policy-check

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1637,40 +1588,108 @@ async fn http_network_or_cache_fetch(
 
-        // Step 3
-        // FIXME: Prompt the user for proxy authentication credentials
+        // FIXME: Step 15.3 If fetchParams is canceled, then return the appropriate network error for fetchParams.
```
Again I think this could be done as mentioned above. TODO for another PR is good too.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1637,40 +1588,108 @@ async fn http_network_or_cache_fetch(
-        // Step 3
-        // FIXME: Prompt the user for proxy authentication credentials
+        // FIXME: Step 15.3 If fetchParams is canceled, then return the appropriate network error for fetchParams.
+        // FIXME: Step 15.4 Prompt the end user as appropriate in request’s window and store the
```
Also more of a TODO(I realize the existing code uses FIXME, but let's clean that up while we're at it).

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1637,40 +1588,108 @@ async fn http_network_or_cache_fetch(
-        // FIXME: Prompt the user for proxy authentication credentials
+        // FIXME: Step 15.3 If fetchParams is canceled, then return the appropriate network error for fetchParams.
+        // FIXME: Step 15.4 Prompt the end user as appropriate in request’s window and store the
+        //                  result as a proxy-authentication entry.
```
By this point I am not so sure about the indenting.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1637,40 +1588,108 @@ async fn http_network_or_cache_fetch(
+}
+
+// TODO: Judging from the name, this appears to be https://fetch.spec.whatwg.org/#cross-origin-resource-policy-check,
+//       but the steps aren't even close to the spec. Perhaps this needs to be rewritten?
```
Yes very good TODO. I think it would be great to have an issue for all those TODOs, with checkboxes.

simonwuelker replied:
I opened #33615 for this (seperate issue since its not directly related to the http network or cache fetch)

gterzian commented on components/script/fetch.rs:
```diff
@@ -146,33 +146,47 @@ pub fn Fetch(
     request_init.csp_list.clone_from(&global.get_csp_list());
 
-    // Step 3
+    // FIXME: Step 4. If requestObject’s signal is aborted, then: [..]
```
TODO. The algo here in script deserves a separate issue for TODOs, please.

gterzian commented on components/shared/net/request.rs:
```diff
@@ -753,3 +859,14 @@ pub fn convert_header_names_to_sorted_lowercase_set(
+        ImmutableOrigin::Opaque(_) => headers::Origin::NULL,
+        ImmutableOrigin::Tuple(scheme, host, port) => {
+            headers::Origin::try_from_parts(scheme, &host.to_string(), *port)
+                .expect("servo and hyper disagree about valid origins")
```
Let's note a TODO and return NULL if this fails.

gterzian commented on components/shared/net/request.rs:
```diff
@@ -753,3 +859,14 @@ pub fn convert_header_names_to_sorted_lowercase_set(
 }
+
+/// <https://html.spec.whatwg.org/multipage/#ascii-serialisation-of-an-origin>
+fn immutable_origin_to_hyper_origin(origin: &ImmutableOrigin) -> headers::Origin {
```
I don't think this needs to be in a function.

gterzian commented on components/shared/net/request.rs:
```diff
@@ -598,6 +599,111 @@ impl Request {
+    }
+
+    /// <https://fetch.spec.whatwg.org/#append-a-request-origin-header>
+    pub fn append_a_request_origin_header(&mut self) {
```
I don't think this belongs here, because it is only used in `components/net`, or do you think other components will need it in the future? If not, this can be a function that takes a request as argument.

gterzian commented on components/shared/net/request.rs:
```diff
@@ -598,6 +599,111 @@ impl Request {
+                        }
+                    },
+                    _ => {
+                        // Do nothing.
```
Otherwise.

gterzian reviewed (changes requested):
Thank you! Couple of points to address, and some issues to open please.

simonwuelker committed:
Move all code into http_loader.rs

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Don't panic if hyper/servo disagree about valid origins

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker commented on components/net/http_loader.rs:
```diff
@@ -1096,25 +1083,32 @@ async fn http_network_or_cache_fetch(
+    // Step 8. Run these steps, but abort when fetchParams is canceled:
     // Step 8.1: If request’s window is "no-window" and request’s redirect mode is "error", then set
-    // httpFetchParams to fetchParams and httpRequest to request.
+    //           httpFetchParams to fetchParams and httpRequest to request.
```
The way I've formatted spec comment is that if they are broken up into multiple lines, then the other lines are indented to start at the same column as the first line. I added whitespace here so its consistent with the other spec comments in the algorithm.

simonwuelker committed:
Add "otherwise" to spec comment

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Remove TODO about No-Store cache directive

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker commented:
> ⚠️ Try run ([#11028869444](https://github.com/servo/servo/actions/runs/11028869444)) failed.

This seems unrelated at first glance and doesn't reproduce on my fork (https://github.com/simonwuelker/servo/actions/runs/11032802482)

jdm commented:
Yeah, I suspect that's an existing intermittent problem.

simonwuelker committed:
Bring http_network_or_cache_fetch closer to the spec

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Convert FIXME's to TODOs when appropriate

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Remove indentation from multiline spec comments

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Add fetch assertions

This is from a spec update where assertions
about requests origin not being client
were added.

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Add note about serializing headers

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Add TODO about partitioning http cache

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker committed:
Convert FIXME to TODO in script/

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker commented:
I opened #33615 and #33616 for the remaining TODOs

gterzian commented:
Thank you, I will review this tomorrow.

gterzian reviewed (approved):
Thank you! As a last note, there are several parts in the code where a "TODO" for stuff you noted as todo in the issue is missing, so would be good to add it in the format `TODO(#{add_issue_number}):` see for example https://github.com/servo/servo/blob/c7a4e4f627f0bf331cedc1700ddeb92ffcba4e78/components/script/script_thread.rs#L1684

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1638,40 +1589,108 @@ async fn http_network_or_cache_fetch(
+    request: &Request,
+    response: &Response,
+) -> CrossOriginResourcePolicy {
+    // Step 1
```
Ah I see this is https://github.com/servo/servo/issues/33615, so forget about it.

simonwuelker committed:
Link to relevant issue for TODO comments

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

simonwuelker commented:
> Thank you! As a last note, there are several parts in the code where a "TODO" for stuff you noted as todo in the issue is missing, so would be good to add it in the format `TODO(#{add_issue_number}):` see for example
> 
> https://github.com/servo/servo/blob/c7a4e4f627f0bf331cedc1700ddeb92ffcba4e78/components/script/script_thread.rs#L1684

Done (:
